### DISCUSSION TITLE
Help with understanding how positions are used to connect nodes to edges

### DISCUSSION THREAD

**Discussion by amorsi1** [amorsi1/JuiJitsu-RL#1] [Q&A]
Hi! @Eelis @ScottyChung can you help me understand some of the details of how the data in transitions correspond to data in positions?

I've processed the GrappleMap.txt file into a dataframe where each row is a position or a transition, and constructed nodes and edges out of that, however when I looked for terminal nodes (meant to signify a submission or the end of a roll) I found that many of them were normal positions that would likely have follow up moves. 

For example: "completed imanari roll" shows up as a terminal node, but on the website it clearly has transitions out of it
<img width="287" alt="Screenshot 2024-08-03 at 2 42 16 PM" src="https://github.com/user-attachments/assets/cf3cdd82-f4a6-441c-a177-648693891a9f">

Looking more closely at that case, it seems like the discrepancy is coming from the fact that I assumed that the 4 lines of the positions data would line up with the first 4 lines of the transition data. From what I can tell it seems like 1) the order of the players encoded position is switched (i.e. the top 2 rows and the bottom 2 rows) and 2) the joint coordinates are also slightly different.

"completed imanari roll" position:

    M4azHhD9azS4JqaFHOBRaEQaKsbHHTCDbGQNLwiyImFhhSSpG8ngK4EXniN5IUwqLPGgv
    EQsKjtBIpHAsZT6NZuUGOKTt3RqOFv3GILAuwQuPYv0GBMOurQfF8rcMwHFwOOjIHzdOX
    GzksJNwtp4TmE3j0M6yuozV6GokvMHyjolUIIQlMS1AQjSP5D9kPX4BBiuWSJ4esXzFXb
    LV4HyhXVECEeKUwF7g2RNDRfQQuE9hlQVEse5PDD1hSRoFLe4PCEPgDZfIxcQVRKudtUe

start position of "to kneebar":

    GzksDFwtp4NeE3j0GYyuozPYGokvGzyjolOAIQlMMTAQjSJXD9kPRWBBiuQKJ4esRrFXb
    LPWHyhXPwCEeKOoF7g2LFDRfQKmE9hlKNEse5JvD1hSLgFLe4JuEPgDS7IxcQPJKudtN6
    M4azA9D9azMWJqaFBGBRaEJ2KsbHBLCDbGKFLwiyCeFhhSMhG8ngEWEXniHXIUwqFHGgv
    EKkKjtBChHAsZNYNZuUAGKTt3LiOFv3AALAuwKmPYv0AtMOurJ7F8rcGoHFwOIbIHzdIP

Do you have any guidance for how I can recognize that the start position of the transition is the same as the position represented in the node? Currently I'm working on decoding the position and trying to tell if the two positions are identical when reoriented, but am not sure that's going to solve the problem.

**Comment by Eelis** [comment]
Yeah, since the simplistic db format currently does not explicitly declare which transitions connect, the code computes the connections by checking if the end of one transition is a reorientation of the start of another transition.

The code for that is [here](https://github.com/Eelis/GrappleMap/blob/master/src/positions.cpp#L198-L259). Basically, it computes the reorientation that would be required to line up the players' heads in the two positions, and then sees if applying that reorientation to the whole position yields the other position.

But this is flaky, so eventually the db should declare the connections explicitly instead.

**Reply by amorsi1** [reply]
Thanks for the quick response! all your hard work on this repo is really impressive - this is by far the best database i could find. I tried to reimplement the C++ code in python but found it a little tricky to reimplement faithfully. Instead I settled on using head2head, checking head to limb distances for both players, then orthogonal procrustes analysis.

I end up with a graph that has 872 nodes and 1508 edges. This is rather different to your 725 nodes and 1485 edges but might be good enough for my purposes. I'm able to save and load my graph as is by saving all the information to a csv then loading from that csv — maybe something similar would work for your db.

I have a handful of questions to ask you about the graph:

**Do you have an idea why I'm getting more edges than you?** I'm defining transitions as any block in the grapplemap.txt file that is greater than 4 lines long, then taking the first 4 and last 4 lines as the start and the end positions. The number of edges I have is completely independent of the position matching, so I'm not sure where the discrepancy is coming from. I noticed there are >100 transitions without a description (the description is '...' ), should those be handled differently?

**What is the logic for what makes a transitions blue/red/black in the explorer?**
![Screenshot 2024-08-15 at 1 01 03 PM](https://github.com/user-attachments/assets/64b274f8-8758-4b96-b8fc-4c2b365e7ab9)

It seems like transitions are either red, blue, or black. How can I tell which one a transition is based on the db? This is really relevant to me because I want to simulate a match, so of course there are certain moves one of the players can do that the other can't

**How can I tell which player is the blue or the red player?** Sometimes the players are swapped in the db (like in the "completed imanari roll" and "to kneebar" which I initially reached out to you about), and I want to make sure it doesn't confuse the game state. I think the blue player would be bummed if he was about to choke out the red player and suddenly their positions were swapped.

**Reply by Eelis** [reply]
The idea is that:
- red is the attacking/top player, and
- blue is the defensive/bottom player.

But if a certain transition is in the db with player 0 on top, then the code understands that the transition is also applicable with player
1 on top. In other words, at all times, the players are interchangable.

The code realizes this by having an extended notion of "reorientation" that includes the possibility of swapping the two players. The code for that is [here](https://github.com/Eelis/GrappleMap/blob/master/src/positions.cpp#L243-L254).

The transitions going out of `completed imanari roll` into `honey hole` or `knee bar` are ones that turn the bottom/defensive player into the top/attacking player. So if your code doesn't incorporate the possibility of this kind of swap, then possibly that's why it's overlooking these transitions.

For transitions, their color is based on whether they have a `top` or `bottom` property in the db. For example, a transition that has:

    properties: bottom

will be colored blue. This indicates that the transition is a move potentially available to the bottom player to improve their position. So on the position page for a given position, red transitions are ones that improve the top player's position, while blue transitions are ones that improve the bottom player's position. (Black transitions are ones I haven't gotten around to annotating with either `top` or `bottom` yet.)

You may be interested in https://eel.is/GrappleMap/transitions.js . It's a conversion of the database to near-JSON, which I used for the web ui. It has all the transitions declared explicitly with `to`/`from` node IDs (and reorientation information).

Edit: Oh and there's nothing special about transitions with a "..." description. It's just a placeholder for a proper name and reference to be figured out later.

**Reply by amorsi1** [reply]
The transitions.js file is exactly what i've been looking for! I've loaded it into a JSON file and constructed my graph out of it (with the correct number of transitions and nodes). Thanks for sending :)

Let me make sure I understand a few things: 

- Does red/blue just indicate which player is on top/attacking? So that if the blue player went from defending to attacking would they become the red player?
- Is the swap_players=True flag in the edge how players switch from between blue and red, or am I missing something?
- Aside from the visual effect, the the red/blue classification is important because it determines what moves they can do based on the transition. Transitions with 'top' or 'bottom' in the properties tag can only be initiated by the corresponding player (i.e. the bottom player can't perform a transition that is labeled 'top')

Is there anything else I’m missing? My short term plan is to just start games at a random, somewhat central position in the graph, and randomly assign players to the top or bottom position. From there, I’ll rely on GrappleMap’s transition logic (by the swap_players flag) to dictate who is on top/bottom and therefore which moves are available to each player.

To create a reward signal for the players I’ll then go through each terminal node in the graph (n=75) and annotate whether it’s a submission and which player wins in that case. I also want to add points from sweeps/takedowns based on some of the move metadata but that seems like it would be a lot more work to go through everything.

**Reply by amorsi1** [reply]
@Eelis Do you have any insight on how symmetrical positions are handled? Is that what the reversible attribute refers to?

Right now I'm initializing my game with a player either on top or on bottom, then that is propagated throughout the rollout. However, encountering a symmetrical position (where there isn't really a top or bottom) is likely to confuse players internal understanding of their relative position and could break my move logic for the rest of the rollout. This can be addressed as long as there are rules for how players lose their top/bottom attributes and for how they regain them after a symmetrical position.
